cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppStateHandler.java

```diff
@@ -0,0 +1,184 @@
+  public void saveState(FullHttpRequest request, HttpResponder responder,
+                        @PathParam("namespace-id") String namespaceId,
+                        @PathParam("app-name") String appName,
+                        @PathParam("state-key") String stateKey) throws Exception {
+    byte[] stateValue = new byte[request.content().readableBytes()];
+    request.content().readBytes(stateValue);
+
+    validateInput(namespaceId, appName, stateKey, stateValue);
+    AppState appStateRequest = new AppState(namespaceId, appName, stateKey, stateValue);
+    applicationLifecycleService.saveState(appStateRequest);
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  /**
+   * Delete a {@link AppState} for a given app-name and state-key.
+   */
+  @DELETE
+  @Path("/namespaces/{namespace-id}/apps/{app-name}/states/{state-key}")
+  public void deleteState(HttpRequest request, HttpResponder responder,
+                          @PathParam("namespace-id") String namespaceId,
+                          @PathParam("app-name") String appName,
+                          @PathParam("state-key") String stateKey) throws Exception {
+    validateInput(namespaceId, appName, stateKey);
+    AppState appStateRequest = new AppState(namespaceId, appName, stateKey);
+    applicationLifecycleService.deleteState(appStateRequest);
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  /**
+   * Delete all {@link AppState} for a given app-name.
+   */
+  @DELETE
+  @Path("/namespaces/{namespace-id}/apps/{app-name}/states")
+  public void deleteAllStates(HttpRequest request, HttpResponder responder,
+                              @PathParam("namespace-id") String namespaceId,
+                              @PathParam("app-name") String appName) throws Exception {
+    validateInput(namespaceId, appName);
+    AppState appStateRequest = new AppState(namespaceId, appName);
+    applicationLifecycleService.deleteAllStates(appStateRequest);
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  private void validateInput(String namespace, String appName) throws Exception {
+    NamespaceId namespaceId = validateNamespace(namespace);
+
+    validateApplication(appName, namespaceId);
+  }
+
+  private void validateInput(String namespace, String appName, String stateKey) throws Exception {
+    validateInput(namespace, appName);
+
+    validateStateKey(stateKey);
+  }
+
+  private void validateInput(String namespace, String appName, String stateKey, byte[] stateValue) throws Exception {
+    validateInput(namespace, appName, stateKey);
+
+    validateStateValue(stateValue);
+  }
+
+  private void validateApplication(String appName, NamespaceId namespaceId)
+    throws BadRequestException, ApplicationNotFoundException {
+    if (appName == null) {
+      throw new BadRequestException("Path parameter app-id cannot be empty");
+    }
+
+    ApplicationId appId = namespaceId.app(appName);
+    applicationLifecycleService.validateApplication(appId);
```

Comment:
it's not strictly correct to do the lookup here in the handler, as it is open to race conditions. For example, the app could exist here, but then right after this method returns, it could get deleted. Then the handler will go and write the state to the table, the app will no longer exist, and we get some orphaned rows.

Instead, the handler should just translate the strings into NamespaceId and validate the appName is a valid one in terms of the character set allowed for apps. The ApplicationLifecycleService methods for saving/retrieving/deleting state should be throwing ApplicationNotFoundException if the app doesn't exist, without the handler needing to check for it.